src/v7.js

```diff
@@ -0,0 +1,55 @@
+import rng from './rng.js';
+import { unsafeStringify } from './stringify.js';
+
+/**
+ * UUID V7 - Unix Epoch time-based UUID
+ *
+ * The IETF has introduced a draft to update RFC4122, introducing 3 new
+ * UUID versions (6,7,8). This implementation of V7 is based on the accepted,
+ * though not yet approved, revisions.
+ *
+ * RFC 4122: https://www.ietf.org/rfc/rfc4122.html
+ * [DRAFT] Updated RFC 4122: https://github.com/ietf-wg-uuidrev/rfc4122bis
+ *
+ * Sample V7 value: https://ietf-wg-uuidrev.github.io/rfc4122bis/draft-00/draft-ietf-uuidrev-rfc4122bis.html#name-example-of-a-uuidv7-value
+ */
+
+function v7(options, buf, offset) {
+  options = options || {};
+
+  // milliseconds since unix epoch, 1970-01-01 00:00
+  const msecs = options.msecs !== undefined ? options.msecs : Date.now();
+
+  // rnds is Uint8Array(16) filled with random bytes
+  const rnds = options.random || (options.rng || rng)();
+
+  // initialize buffer and pointer
+  let i = (buf && offset) || 0;
+  const b = buf || new Uint8Array(16);
+
+  // [bytes 0-5] 48 bits of timestamp
+  b[i++] = (msecs / 0x10000000000) & 0xff;
+  b[i++] = (msecs / 0x100000000) & 0xff;
+  b[i++] = (msecs / 0x1000000) & 0xff;
+  b[i++] = (msecs / 0x10000) & 0xff;
+  b[i++] = (msecs / 0x100) & 0xff;
+  b[i++] = msecs & 0xff;
+
+  // [byte 6] - set 4 bits of version (7)
+  b[i++] = (rnds[6] & 0x0f) | 0x70;
+
+  // [byte 7] - set 8 bits of random
+  b[i++] = rnds[7];
+
+  // [byte 8] - Per RFC4122 4.1.1, set the variant (2 bits)
+  b[i++] = (rnds[8] & 0x3f) | 0x80;
+
+  // [bytes 9-16] populate with random bytes
+  for (let n = 0; n < 7; ++n) {
+    b[i + n] = rnds[n + 9];
+  }
```

Comment:
Not sure if we want to change anything here, but unrolling the loop is _potentially_ faster and _potentially_ smaller when compressed with gzip.

```suggestion
  b[i++] = rnds[9];
  b[i++] = rnds[10];
  b[i++] = rnds[11];
  b[i++] = rnds[12];
  b[i++] = rnds[13];
  b[i++] = rnds[14];
  b[i++] = rnds[15];
```

note that V8 doesn't do loop unrolling: https://stackoverflow.com/a/66082249/148072

My guess for why it might be smaller when gzipped is because `];  b[i++] = rnds` will be repeated a lot, which should compress very well.

I don't think that it's a prerequisite for landing this, but I think it would be nice to benchmark and compare compressed size! If/when I get some time over I shall do this, since I think it's very interesting!

Reply:
(a quick test of pasting just this function into [this online gzip compressor](https://facia.dev/tools/compress-decompress/gzip-compress/) gives 652B for the loop, and 640B when it's unrolled. This isn't running on the minified code though so is only indicative)

Reply:
A quick test seems to suggest that it's about 2.5% faster to unroll the loop in Chrome 109.0.5414.87 (Official Build) (arm64), and about the same (but the unrolled seems to always win, even if with a 0% margin) in Safari Version 16.2 (18614.3.7.1.5) (arm64).

https://jsbench.me/9vldb9u08i/1

Reply:
> However in this case I kind of like how the unrolled code reads. It flows well with the other `b[i++]` assignments in prior bytes, so I'm good with either form.

Yeah I agree, that was actually what got me started on this, hehe 😁

Reply:
@LinusU I agree, the unrolled code reads more clearly and fits cleanly with the other byte manipulations. As an aside, I enjoyed going through your performance breakdown - learned a few things!